rundeckapp/grails-app/services/rundeck/services/ScheduledExecutionService.groovy

```diff
@@ -4221,17 +4222,64 @@ class ScheduledExecutionService implements ApplicationContextAware, Initializing
+                    eq('serverNodeUUID', fromServerUUID)
+                } else {
+                    isNull('serverNodeUUID')
+                }
+            } else {
+                or {
+                    isNull('serverNodeUUID')
+                    ne('serverNodeUUID', toServerUUID)
+                }
+            }
+        }
+    }
+    /**
+     * Return jobs  that should be claimed for rescheduling on a cluster member, which includes
+     * all jobs that are have adhoc scheduled executions, where the cluster owner
+     * matches the criteria
+     * @param toServerUUID node ID of claimant
+     * @param fromServerUUID null for unclaimed jobs, or owner server node ID
+     * @param selectAll if true, match jobs not owned by claimant, if false match based on fromServerUUID setting
+     * @param projectFilter project name for jobs, or null for any project
+     * @return
+     */
+    List<ScheduledExecution> getJobsWithAdhocScheduledExecutionsToClaim(String toServerUUID, String fromServerUUID, boolean selectAll, String projectFilter) {
+        return ScheduledExecution.createCriteria().listDistinct {
+            applyAdhocScheduledExecutionsCriteria(delegate, selectAll, fromServerUUID, toServerUUID, projectFilter)
+        }
+    }
+        /**
+     * Return jobs  that should be claimed for rescheduling on a cluster member, which includes
+     * all jobs that are either *scheduled* or in the jobids list or have adhoc scheduled executions, where the cluster owner
+     * matches the criteria
+     * @param toServerUUID node ID of claimant
+     * @param fromServerUUID null for unclaimed jobs, or owner server node ID
+     * @param selectAll if true, match jobs not owned by claimant, if false match based on fromServerUUID setting
+     * @param projectFilter project name for jobs, or null for any project
+     * @param jobids explicit list of job IDs to select
+     * @param ignoreInnerScheduled if true, do not select only scheduled jobs
+     * @return
+     */
+    List<ScheduledExecution> getSchedulesJobToClaim(String toServerUUID, String fromServerUUID, boolean selectAll, String projectFilter, List<String> jobids, ignoreInnerScheduled = false) {
 
-    List getSchedulesJobToClaim(String toServerUUID, String fromServerUUID, boolean selectAll, String projectFilter, List<String> jobids, ignoreInnerScheduled = false) {
-        def c = ScheduledExecution.createCriteria()
-        def scheduledExecutions = c.listDistinct {
+        return ScheduledExecution.createCriteria().listDistinct {
             or {
-                if(!ignoreInnerScheduled){
-                    eq('scheduled', true)
-                }
-                executions(CriteriaSpecification.LEFT_JOIN) {
-                    eq('status', ExecutionService.EXECUTION_SCHEDULED)
-                    isNull('dateCompleted')
+                applyAdhocScheduledExecutionsCriteria(delegate, selectAll, fromServerUUID, toServerUUID, projectFilter)
```

Comment:
This line makes the query slower by orders of magnitude. We might want to find a way to optimize this.

Reply:
hmm, i just moved the `executions` join construction to a separate method call, that shouldn't make it any slower.  do you mean the restructuring of the query made it slower? this is possible but unfortunately the previous query was not correct for the results we are looking for.

Reply:
Maybe it was slow to begin with. If I comment out this line and run the api takeover the query time goes from 80 seconds to about 1 second.